src/game/client/shadow/w3dbuffermanager.cpp

```diff
@@ -0,0 +1,475 @@
+        buf->m_usedSlots = slot;
+        slot->m_size = size;
+        slot->m_start = 0;
+        slot->m_VB = buf;
+        slot->m_nextSameVB = nullptr;
+        slot->m_prevSameVB = nullptr;
+        slot->m_nextSameSize = nullptr;
+        slot->m_prevSameSize = nullptr;
+
+        return slot;
+    }
+
+    return nullptr;
+}
+
+W3DBufferManager::W3DIndexBufferSlot *W3DBufferManager::Get_Slot(int size)
+{
+    size = (size + 31) & ~31;
+    int size_index = (size / 32) - 1;
+
+    captainslog_dbgassert(size_index < MAX_IB_SIZES && size, "Allocating too large index buffer slot");
+
+    W3DIndexBufferSlot *ib_slot = m_W3DIndexBufferSlots[size_index];
+
+    if (ib_slot) {
+        m_W3DIndexBufferSlots[size_index] = ib_slot->m_nextSameSize;
+        if (ib_slot->m_nextSameSize) {
+            ib_slot->m_nextSameSize->m_prevSameSize = nullptr;
+        }
+
+        return ib_slot;
+    }
+
+    return Allocate_Slot_Storage(size);
+}
+
+void W3DBufferManager::Release_Slot(W3DIndexBufferSlot *ib_slot)
+{
+    int size_index = (ib_slot->m_size / 32) - 1;
+
+    ib_slot->m_nextSameSize = m_W3DIndexBufferSlots[size_index];
+
+    if (m_W3DIndexBufferSlots[size_index]) {
+        m_W3DIndexBufferSlots[size_index]->m_prevSameSize = ib_slot;
+    }
+
+    m_W3DIndexBufferSlots[size_index] = ib_slot;
+}
+
+W3DBufferManager::W3DIndexBufferSlot *W3DBufferManager::Allocate_Slot_Storage(int size)
+{
+    captainslog_dbgassert(m_numEmptyIndexSlotsAllocated < MAX_NUMBER_SLOTS, "Nore more IB Slots");
+
+    for (W3DIndexBuffer *ib = m_W3DIndexBuffers; ib; ib = ib->m_nextIB) {
+
+        if (ib->m_size - ib->m_startFreeIndex >= size && m_numEmptyIndexSlotsAllocated < MAX_NUMBER_SLOTS) {
+            W3DIndexBufferSlot *slot = &m_W3DIndexBufferEmptySlots[m_numEmptyIndexSlotsAllocated];
+
+            slot->m_size = size;
+            slot->m_start = ib->m_startFreeIndex;
+            slot->m_IB = ib;
+            slot->m_nextSameIB = ib->m_usedSlots;
+            slot->m_prevSameIB = nullptr;
+
+            if (ib->m_usedSlots) {
+                ib->m_usedSlots->m_prevSameIB = slot;
+            }
+
+            slot->m_nextSameSize = nullptr;
+            slot->m_prevSameSize = nullptr;
+            ib->m_usedSlots = slot;
+            ib->m_startFreeIndex += size;
+            ++m_numEmptyIndexSlotsAllocated;
+
+            return slot;
+        }
+    }
+
+    W3DIndexBuffer *next = m_W3DIndexBuffers;
+
+    captainslog_dbgassert(
+        m_numEmptyIndexBuffersAllocated < MAX_INDEX_BUFFERS_CREATED, "Reached Max Static IB Shadow Geometry");
+
+    if (m_numEmptyIndexBuffersAllocated < MAX_INDEX_BUFFERS_CREATED) {
+        m_W3DIndexBuffers = &m_W3DEmptyIndexBuffers[m_numEmptyIndexBuffersAllocated];
+        m_W3DIndexBuffers->m_nextIB = next;
+        ++m_numEmptyIndexBuffersAllocated;
+
+        W3DIndexBuffer *buf = m_W3DIndexBuffers;
+
+        int count;
```

Comment:
You can combine these 5 lines in one with `? :` operator. Or use `std::max(32768, size)`.